cli/cli_app.go

```diff
@@ -261,7 +302,37 @@ func runTerragruntWithConfig(terragruntOptions *options.TerragruntOptions, terra
 			return err
 		}
 	}
-	return shell.RunTerraformCommand(terragruntOptions, terragruntOptions.TerraformCliArgs...)
+
+	terragruntOptions.Logger.Println("Running terraform with: %s", terragruntOptions)
+
+	errorOccurred := true
+
+	if len(terragruntConfig.Terraform.BeforeHook) > 0 {
+		terragruntOptions.Logger.Printf("Detected %d Before Hooks", len(terragruntConfig.Terraform.BeforeHook))
+		errorOccurred = processBeforeAfterHookActions(terragruntConfig.Terraform.BeforeHook, terragruntConfig, terragruntOptions, false)
+	}
+
+	var possibleError error
+
+	if !errorOccurred {
+		possibleError = shell.RunTerraformCommand(terragruntOptions, terragruntOptions.TerraformCliArgs...)
+	} else {
+		terragruntOptions.Logger.Printf("Previous error detected. Not running terragrunt")
+	}
+
+	if len(terragruntConfig.Terraform.AfterHook) > 0 {
+		terragruntOptions.Logger.Printf("Detected %d After Hooks", len(terragruntConfig.Terraform.AfterHook))
+
+		if errorOccurred || possibleError != nil {
+			//if there was an error during before hook run
+			//or if there was an error during terraform execution
+			//make sure to pass that knowledge on to after hook execution
+			processBeforeAfterHookActions(terragruntConfig.Terraform.AfterHook, terragruntConfig, terragruntOptions, true)
+		} else {
+			processBeforeAfterHookActions(terragruntConfig.Terraform.AfterHook, terragruntConfig, terragruntOptions, false)
+		}
+	}
+	return possibleError
```

Comment:
It took a little while to figure out the intent of the code above. Some ideas on how to make this a bit clearer.

First, as this method can now generate multiple errors, create a type to capture them:

```go
type MultiError struct {
  Errors []error
}

func NewMultiError(errs ... error) error {
  if len(errs) == 0 {
    return nil
  }

  return MultiError{Errors: errs}
} 

func (errs MultiError) Error() string {
  errorMessages := []string{}
  for _, err := range errs {
    errorMessages = append (out, err.Error())
  }

  return fmt.Sprintf("Hit multiple errors:\n%s", strings.Join(errorMessages, "\n"))
}
```

Next, extract most of this functionality into small helper methods, and collect the errors as you go:

```go
beforeHooksErr := processHooks(terragruntConfig.Terraform.BeforeHooks, terragruntOptions)
terraformCmdErr := runTerraformCommandIfNoErrors(beforeHooksErr, terragruntOptions)
afterHooksErr := processHooks(terragruntConfig.Terraform.afterHooks, terragruntOptions, beforeHooksErr, terraformCmdErr)

return NewMultiError(beforeHooksErr, terraformCmdErr, afterHooksErr)
```

The key things to notice are:

1. `processHooks` takes an optional `...` of errors that happened previous as the final params. On the first call, this list is empty; on the second call, it has errors from before hooks and the Terraform command execution. If any of these are non-nil, it internally checks the `RunOnError` setting to determine if the hook should be executed anyway.

1. The `runTerraformCommandIfNoErrors` only executes if the errors passed into it are `nil`.